include/envoy/upstream/cluster_manager.h

```diff
@@ -22,6 +22,38 @@
 namespace Envoy {
 namespace Upstream {
 
+/**
+ * ClusterUpdateCallbacks provide a way to exposes Cluster lifecycle events in the
+ * ClusterManager.
+ */
+class ClusterUpdateCallbacks {
+public:
+  virtual ~ClusterUpdateCallbacks() {}
+
+  /**
+   * onClusterAddOrUpdate is called when a new cluster is added or an existing cluster
+   * is updated in the ClusterManager.
+   * @param cluster is the ThreadLocalCluster that represents the updated
+   * cluster.
+   */
+  virtual void onClusterAddOrUpdate(ThreadLocalCluster& cluster) PURE;
+
+  /**
+   * onClusterRemoval is called when a cluster is removed; the argument is the cluster name.
+   * @param cluster_name is the name of the removed cluster.
+   */
+  virtual void onClusterRemoval(const std::string& cluster_name) PURE;
+};
+
+/**
+ * CallbackRegistration is a RAII wrapper for a ClusterUpdateCallbacks. Deleting
+ * the CallbackRegistration will remove the callbacks from ClusterManager in O(1).
+ */
+class CallbackRegistration {
+public:
+  virtual ~CallbackRegistration() {}
+};
+
```

Comment:
What about `source/common/common/callback_impl.h`? Could you use this here? If not, I suggest calling this `ClusterUpdateCallbackHandle` and also doing a type alias for the  `std::unique_ptr` variant to `ClusterUpdateCallbackHandlePtr`.

Reply:
Done: renamed to ClusterUpdateCallbacksHandle and created ClusterUpdateCallbacksHandlePtr.